Biometric auth mysteriously fails after adding new fingerprint (UX issue)

## Bug description
In the new code (thanks @nindanaoto) per https://github.com/connectbot/connectbot/pull/1676 to add biometric key support, the following is set, which is also apparently the default:
```
            // Invalidate key if new biometrics are enrolled
            setInvalidatedByBiometricEnrollment(true)
```
However, there's not a proper error message shown when this situation arises. And, perhaps a note about the same should be shown when generating the key, so that people don't inadvertently get locked out of their servers (didn't happen to me, but I can see it happening).

I see this in the PR:
```
	<string name="pubkey_biometric_key_invalidated">Key has been invalidated. This can happen when new fingerprints are enrolled. Please generate a new key.</string>
```

However, it is never shown. See actual exception per logcat in the "Additional Context" section.

## Steps to reproduce
1. Add biometric key in ConnectBot and on the remote server
2. Verify key works
3. Enroll new fingerprint
4. Verify that the key no longer works, and there's not a clear reason given for why this is ("Connection Lost")

## Expected behavior
An error message is shown describing that the previous key was invalidated due to enrolling a new fingerprint.

## Screenshots

<img width="540" height="1205" alt="Image" src="https://github.com/user-attachments/assets/8718a8ad-6907-4e5b-a8a2-56900b74e6f4" />

## Android device

 - Device: Pixel 10 Pro
 - OS: Android 16
 - ConnectBot Version: ConnectBot-git-v1.9.13-296-ge478c655-google.apk

## Server information

 - OS: Debian 12
 - SSH Software and Version: OpenSSH 1:9.2p1-2+deb12u7
 - Pubkeys used (if applicable): ed25519 w/RSA (biometric) private key

## Additional context
```
01-11 02:43:46.734  1861  5165 D CoreBackPreview: Window{c62489 u0 BiometricPrompt}: Setting back callback null
01-11 02:43:46.736   538   571 I keystore2: system/security/keystore2/src/authorization.rs:116 - add_auth_token(challenge=0, userId=censored authId=-censored, authType=0x2, timestamp=439789920ms)
01-11 02:43:46.737  1861  2893 D BiometricService/AuthSession: addAuthToken: 0
01-11 02:43:46.738  1861  2893 D BiometricService: resetLockout from: censored, for: 1, userId: 0
01-11 02:43:46.740  1861  2893 D BiometricService/AuthSession: Cancelling sensorId: censored
01-11 02:43:46.740  1861  2893 D FingerprintService: cancelAuthenticationFromService, sensorId: censored
01-11 02:43:46.741  1861  2951 D BiometricScheduler: [Added] {[335] com.android.server.biometrics.sensors.face.aidl.FaceResetLockoutClient, proto=12, owner=android, cookie=0, requestId=-1, userId=0}, new queue size: 1
01-11 02:43:46.741  1861  2951 D BiometricScheduler: [Polled] {[335] com.android.server.biometrics.sensors.face.aidl.FaceResetLockoutClient, proto=12, owner=android, cookie=0, requestId=-1, userId=0}, State: 0
01-11 02:43:46.742  1861  2949 D BiometricScheduler: cancelAuthenticationOrDetection, requestId: 7
01-11 02:43:46.742  1861  2951 D BiometricScheduler: [Started] {[335] com.android.server.biometrics.sensors.face.aidl.FaceResetLockoutClient, proto=12, owner=android, cookie=0, requestId=-1, userId=0}
01-11 02:43:46.747  1572  1728 I android.hardware.biometrics.face-service.pixel: [ThreadPool]resetLockout(): Operation scheduled.
01-11 02:43:46.749   538   571 E keystore2: system/security/keystore2/src/error.rs:184 - system/security/keystore2/src/security_level.rs:327
01-11 02:43:46.749   538   571 E keystore2: 
01-11 02:43:46.749   538   571 E keystore2: Caused by:
01-11 02:43:46.749   538   571 E keystore2:     0: system/security/keystore2/src/enforcements.rs:586: No suitable auth token for [SecureUserId(-censored)] type Some(r#FINGERPRINT) received in last 30s found
01-11 02:43:46.749   538   571 E keystore2:     1: Error::Km(r#KEY_USER_NOT_AUTHENTICATED)
01-11 02:43:46.749 29816 29838 W System.err: java.io.IOException: android.security.keystore.UserNotAuthenticatedException: User not authenticated
01-11 02:43:46.749  1572 30735 I android.hardware.biometrics.face-service.pixel: Task started: resetLockout{154}
01-11 02:43:46.749  1861  2700 I ImeTracker: system_server:f7c77a0: onRequestHide at ORIGIN_SERVER reason IME_REQUESTED_CHANGED_LISTENER fromUser false
01-11 02:43:46.750 29816 29838 W System.err: 	at com.trilead.ssh2.signature.RSASHA1Verify.generateSignature(r8-map-id-cb92c09615f8c2214099bfc0686d199521b1a094e6bd2daf717207535a8c5dee:158)
01-11 02:43:46.750 29816 29838 W System.err: 	at androidx.compose.runtime.ProvidedValue.authenticatePublicKey(r8-map-id-cb92c09615f8c2214099bfc0686d199521b1a094e6bd2daf717207535a8c5dee:139)
01-11 02:43:46.750 29816 29838 W System.err: 	at com.trilead.ssh2.Connection.authenticateWithPublicKey(r8-map-id-cb92c09615f8c2214099bfc0686d199521b1a094e6bd2daf717207535a8c5dee:22)
01-11 02:43:46.750 29816 29838 W System.err: 	at org.connectbot.transport.SSH.tryPublicKey(r8-map-id-cb92c09615f8c2214099bfc0686d199521b1a094e6bd2daf717207535a8c5dee:10)
01-11 02:43:46.750 29816 29838 W System.err: 	at org.connectbot.transport.SSH.tryPublicKey(r8-map-id-cb92c09615f8c2214099bfc0686d199521b1a094e6bd2daf717207535a8c5dee:9)
01-11 02:43:46.750 29816 29838 W System.err: 	at org.connectbot.transport.SSH.authenticate(r8-map-id-cb92c09615f8c2214099bfc0686d199521b1a094e6bd2daf717207535a8c5dee:359)
01-11 02:43:46.750 29816 29838 W System.err: 	at org.connectbot.transport.SSH.connect(r8-map-id-cb92c09615f8c2214099bfc0686d199521b1a094e6bd2daf717207535a8c5dee:375)
01-11 02:43:46.750 29816 29838 W System.err: 	at org.connectbot.service.TerminalBridge$startConnection$2.invokeSuspend(r8-map-id-cb92c09615f8c2214099bfc0686d199521b1a094e6bd2daf717207535a8c5dee:161)
01-11 02:43:46.750 29816 29838 W System.err: 	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(r8-map-id-cb92c09615f8c2214099bfc0686d199521b1a094e6bd2daf717207535a8c5dee:9)
01-11 02:43:46.750 29816 29838 W System.err: 	at kotlinx.coroutines.DispatchedTask.run(r8-map-id-cb92c09615f8c2214099bfc0686d199521b1a094e6bd2daf717207535a8c5dee:112)
01-11 02:43:46.750 29816 29838 W System.err: 	at androidx.core.app.ActivityRecreator$1.run(r8-map-id-cb92c09615f8c2214099bfc0686d199521b1a094e6bd2daf717207535a8c5dee:19)
01-11 02:43:46.750 29816 29838 W System.err: 	at kotlinx.coroutines.scheduling.TaskImpl.run(r8-map-id-cb92c09615f8c2214099bfc0686d199521b1a094e6bd2daf717207535a8c5dee:3)
01-11 02:43:46.750 29816 29838 W System.err: 	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(r8-map-id-cb92c09615f8c2214099bfc0686d199521b1a094e6bd2daf717207535a8c5dee:87)
01-11 02:43:46.750 29816 29838 W System.err: Caused by: android.security.keystore.UserNotAuthenticatedException: User not authenticated
01-11 02:43:46.750 29816 29838 W System.err: 	at android.security.keystore2.KeyStoreCryptoOperationUtils.getInvalidKeyException(KeyStoreCryptoOperationUtils.java:125)
01-11 02:43:46.750 29816 29838 W System.err: 	at android.security.keystore2.AndroidKeyStoreSignatureSpiBase.ensureKeystoreOperationInitialized(AndroidKeyStoreSignatureSpiBase.java:217)
01-11 02:43:46.750 29816 29838 W System.err: 	at android.security.keystore2.AndroidKeyStoreSignatureSpiBase.engineInitSign(AndroidKeyStoreSignatureSpiBase.java:123)
01-11 02:43:46.750 29816 29838 W System.err: 	at java.security.Signature$Delegate.init(Signature.java:1364)
01-11 02:43:46.750 29816 29838 W System.err: 	at java.security.Signature$Delegate.chooseProvider(Signature.java:1314)
01-11 02:43:46.750 29816 29838 W System.err: 	at java.security.Signature$Delegate.engineInitSign(Signature.java:1400)
01-11 02:43:46.750 29816 29838 W System.err: 	at java.security.Signature.initSign(Signature.java:710)
01-11 02:43:46.750 29816 29838 W System.err: 	at com.trilead.ssh2.signature.RSASHA1Verify.generateSignature(r8-map-id-cb92c09615f8c2214099bfc0686d199521b1a094e6bd2daf717207535a8c5dee:136)
01-11 02:43:46.750 29816 29838 W System.err: 	... 12 more
01-11 02:43:46.755  1861  1916 W WindowManager: container Window{c62489 u0 BiometricPrompt}does not have a surface
01-11 02:43:46.756  1572  2956 I android.hardware.biometrics.face-service.pixel: [ThreadPool]resetLockout(): Operation started.
01-11 02:43:46.757  1572  2956 I android.hardware.biometrics.face-service.pixel: Task finished: resetLockout{154}
01-11 02:43:46.759  1572  2956 I android.hardware.biometrics.face-service.pixel: [ThreadPool]resetLockout(): Operation finished.
```

--- a/app/src/main/java/org/connectbot/transport/SSH.kt
+++ b/app/src/main/java/org/connectbot/transport/SSH.kt
@@ -19,7 +19,8 @@ package org.connectbot.transport
 
 import android.content.Context
 import android.net.Uri
-import timber.log.Timber
+import android.security.keystore.KeyPermanentlyInvalidatedException
+import androidx.core.net.toUri
 import com.trilead.ssh2.AuthAgentCallback
 import com.trilead.ssh2.ChannelCondition
 import com.trilead.ssh2.Connection
@@ -52,6 +53,7 @@ import org.connectbot.service.requestHostKeyFingerprintPrompt
 import org.connectbot.service.requestStringPrompt
 import org.connectbot.util.HostConstants
 import org.connectbot.util.PubkeyUtils
+import timber.log.Timber
 import java.io.IOException
 import java.io.InputStream
 import java.io.OutputStream
@@ -71,18 +73,24 @@ import java.security.interfaces.RSAPublicKey
 import java.security.spec.InvalidKeySpecException
 import java.util.Locale
 import java.util.regex.Pattern
-import androidx.core.net.toUri
 
 /**
  * @author Kenny Root
  */
-class SSH : AbsTransport, ConnectionMonitor, InteractiveCallback, AuthAgentCallback {
+class SSH :
+    AbsTransport,
+    ConnectionMonitor,
+    InteractiveCallback,
+    AuthAgentCallback {
 
     private var compression = false
+
     @Volatile
     private var authenticated = false
+
     @Volatile
     private var connected = false
+
     @Volatile
     private var sessionOpen = false
 
@@ -112,45 +120,39 @@ class SSH : AbsTransport, ConnectionMonitor, InteractiveCallback, AuthAgentCallb
 
     constructor(host: Host?, bridge: TerminalBridge?, manager: TerminalManager?) : super(host, bridge, manager)
 
-    private fun decodePublicKey(algorithm: String, keyBlob: ByteArray): PublicKey? {
-        return try {
-            when (algorithm) {
-                "ssh-rsa" -> RSASHA1Verify.get().decodePublicKey(keyBlob)
-                "ssh-dss" -> DSASHA1Verify.get().decodePublicKey(keyBlob)
-                "ssh-ed25519" -> Ed25519Verify.get().decodePublicKey(keyBlob)
-                "ecdsa-sha2-nistp256" -> ECDSASHA2Verify.ECDSASHA2NISTP256Verify.get().decodePublicKey(keyBlob)
-                "ecdsa-sha2-nistp384" -> ECDSASHA2Verify.ECDSASHA2NISTP384Verify.get().decodePublicKey(keyBlob)
-                "ecdsa-sha2-nistp521" -> ECDSASHA2Verify.ECDSASHA2NISTP521Verify.get().decodePublicKey(keyBlob)
-                else -> null
-            }
-        } catch (e: IOException) {
-            Timber.e(e, "Failed to decode public key")
-            null
+    private fun decodePublicKey(algorithm: String, keyBlob: ByteArray): PublicKey? = try {
+        when (algorithm) {
+            "ssh-rsa" -> RSASHA1Verify.get().decodePublicKey(keyBlob)
+            "ssh-dss" -> DSASHA1Verify.get().decodePublicKey(keyBlob)
+            "ssh-ed25519" -> Ed25519Verify.get().decodePublicKey(keyBlob)
+            "ecdsa-sha2-nistp256" -> ECDSASHA2Verify.ECDSASHA2NISTP256Verify.get().decodePublicKey(keyBlob)
+            "ecdsa-sha2-nistp384" -> ECDSASHA2Verify.ECDSASHA2NISTP384Verify.get().decodePublicKey(keyBlob)
+            "ecdsa-sha2-nistp521" -> ECDSASHA2Verify.ECDSASHA2NISTP521Verify.get().decodePublicKey(keyBlob)
+            else -> null
         }
+    } catch (e: IOException) {
+        Timber.e(e, "Failed to decode public key")
+        null
     }
 
-    private fun getKeySize(publicKey: PublicKey?): Int {
-        return when (publicKey) {
-            is RSAPublicKey -> publicKey.modulus.bitLength()
-            is DSAPublicKey -> publicKey.params.p.bitLength()
-            is ECPublicKey -> publicKey.params.curve.field.fieldSize
-            is Ed25519PublicKey -> 256
-            else -> 0
-        }
+    private fun getKeySize(publicKey: PublicKey?): Int = when (publicKey) {
+        is RSAPublicKey -> publicKey.modulus.bitLength()
+        is DSAPublicKey -> publicKey.params.p.bitLength()
+        is ECPublicKey -> publicKey.params.curve.field.fieldSize
+        is Ed25519PublicKey -> 256
+        else -> 0
     }
 
-    private fun getKeyType(openSshKeyType: String): String? {
-        return if (openSshKeyType == "ssh-rsa") {
-            "RSA"
-        } else if (openSshKeyType == "ssh-dss") {
-            "DSA"
-        } else if (openSshKeyType == "ssh-ed25519") {
-            "Ed25519"
-        } else if (openSshKeyType.startsWith("ecdsa-sha2-")) {
-            "EC"
-        } else {
-            null
-        }
+    private fun getKeyType(openSshKeyType: String): String? = if (openSshKeyType == "ssh-rsa") {
+        "RSA"
+    } else if (openSshKeyType == "ssh-dss") {
+        "DSA"
+    } else if (openSshKeyType == "ssh-ed25519") {
+        "Ed25519"
+    } else if (openSshKeyType.startsWith("ecdsa-sha2-")) {
+        "EC"
+    } else {
+        null
     }
 
     inner class HostKeyVerifier : ExtendedServerHostKeyVerifier() {
@@ -293,10 +295,8 @@ class SSH : AbsTransport, ConnectionMonitor, InteractiveCallback, AuthAgentCallb
             }
         }
 
-        override fun getKnownKeyAlgorithmsForHost(host: String, port: Int): List<String>? {
-            return this@SSH.host?.id?.let { hostId ->
-                manager?.hostRepository?.getHostKeyAlgorithmsForHostBlocking(hostId)
-            }
+        override fun getKnownKeyAlgorithmsForHost(host: String, port: Int): List<String>? = this@SSH.host?.id?.let { hostId ->
+            manager?.hostRepository?.getHostKeyAlgorithmsForHostBlocking(hostId)
         }
 
         override fun removeServerHostKey(host: String, port: Int, algorithm: String, hostKey: ByteArray) {
@@ -347,7 +347,6 @@ class SSH : AbsTransport, ConnectionMonitor, InteractiveCallback, AuthAgentCallb
                 pubkeyId != HostConstants.PUBKEYID_NEVER &&
                 connection?.isAuthMethodAvailable(currentHost.username, AUTH_PUBLICKEY) == true
             ) {
-
                 // if explicit pubkey defined for this host, then prompt for password as needed
                 // otherwise just try all in-memory keys held in terminalmanager
 
@@ -355,8 +354,9 @@ class SSH : AbsTransport, ConnectionMonitor, InteractiveCallback, AuthAgentCallb
                     // try each of the in-memory keys
                     bridge?.outputLine(manager?.res?.getString(R.string.terminal_auth_pubkey_any))
                     manager?.loadedKeypairs?.entries?.forEach { entry ->
-                        if (entry.value.pubkey?.confirmation == true && !promptForPubkeyUse(entry.key))
+                        if (entry.value.pubkey?.confirmation == true && !promptForPubkeyUse(entry.key)) {
                             return@forEach
+                        }
 
                         val keyPair = entry.value.pair ?: return@forEach
 
@@ -370,10 +370,11 @@ class SSH : AbsTransport, ConnectionMonitor, InteractiveCallback, AuthAgentCallb
                     // use a specific key for this host, as requested
                     val pubkey = manager?.pubkeyRepository?.getByIdBlocking(pubkeyId)
 
-                    if (pubkey == null)
+                    if (pubkey == null) {
                         bridge?.outputLine(manager?.res?.getString(R.string.terminal_auth_pubkey_invalid))
-                    else if (tryPublicKey(pubkey))
+                    } else if (tryPublicKey(pubkey)) {
                         finishConnection()
+                    }
                 }
 
                 pubkeysExhausted = true
@@ -421,8 +422,9 @@ class SSH : AbsTransport, ConnectionMonitor, InteractiveCallback, AuthAgentCallb
     @Throws(NoSuchAlgorithmException::class, InvalidKeySpecException::class, IOException::class)
     private fun tryPublicKey(pubkey: Pubkey): Boolean {
         if (pubkey.confirmation && manager?.isKeyLoaded(pubkey.nickname) == true) {
-            if (!promptForPubkeyUse(pubkey.nickname))
+            if (!promptForPubkeyUse(pubkey.nickname)) {
                 return false
+            }
         }
 
         val pair = getOrUnlockKey(pubkey) ?: return false
@@ -483,6 +485,12 @@ class SSH : AbsTransport, ConnectionMonitor, InteractiveCallback, AuthAgentCallb
                 manager?.addBiometricKey(pubkey, keystoreAlias, publicKey)
                 Timber.d(String.format("Unlocked biometric key '%s'", pubkey.nickname))
                 pair
+            } catch (e: KeyPermanentlyInvalidatedException) {
+                val message = manager?.res?.getString(R.string.terminal_auth_biometric_invalidated, pubkey.nickname)
+                    ?: String.format("Biometric key '%s' has been invalidated. Please generate a new key.", pubkey.nickname)
+                Timber.e(e, message)
+                bridge?.outputLine(message)
+                null
             } catch (e: Exception) {
                 val message = String.format("Failed to load biometric key '%s': %s", pubkey.nickname, e.message)
                 Timber.e(e, message)
@@ -501,8 +509,9 @@ class SSH : AbsTransport, ConnectionMonitor, InteractiveCallback, AuthAgentCallb
             )
 
             // Something must have interrupted the prompt.
-            if (password == null)
+            if (password == null) {
                 return null
+            }
         }
 
         val pair = if (pubkey.type == "IMPORTED") {
@@ -547,8 +556,9 @@ class SSH : AbsTransport, ConnectionMonitor, InteractiveCallback, AuthAgentCallb
     @Throws(IOException::class)
     private fun tryPublicKey(username: String, keyNickname: String, pair: KeyPair): Boolean {
         val success = connection?.authenticateWithPublicKey(username, pair) == true
-        if (!success)
+        if (!success) {
             bridge?.outputLine(manager?.res?.getString(R.string.terminal_auth_pubkey_fail, keyNickname))
+        }
         return success
     }
 
@@ -578,8 +588,9 @@ class SSH : AbsTransport, ConnectionMonitor, InteractiveCallback, AuthAgentCallb
         try {
             session = connection?.openSession()
 
-            if (useAuthAgent != HostConstants.AUTHAGENT_NO)
+            if (useAuthAgent != HostConstants.AUTHAGENT_NO) {
                 session?.requestAuthAgentForwarding(this)
+            }
 
             session?.requestPTY(getEmulation(), columns, rows, width, height, null)
             session?.startShell()
@@ -834,8 +845,9 @@ class SSH : AbsTransport, ConnectionMonitor, InteractiveCallback, AuthAgentCallb
                 val message = t.message
                 if (message != null) {
                     bridge?.outputLine(message)
-                    if (t is NoRouteToHostException)
+                    if (t is NoRouteToHostException) {
                         bridge?.outputLine(manager?.res?.getString(R.string.terminal_no_route))
+                    }
                 }
                 t = t.cause
             }
@@ -931,13 +943,12 @@ class SSH : AbsTransport, ConnectionMonitor, InteractiveCallback, AuthAgentCallb
         stdin?.write(c)
     }
 
-    override fun getOptions(): Map<String, String> {
-        return mapOf("compression" to compression.toString())
-    }
+    override fun getOptions(): Map<String, String> = mapOf("compression" to compression.toString())
 
     override fun setOptions(options: Map<String, String>) {
-        if (options.containsKey("compression"))
+        if (options.containsKey("compression")) {
             compression = options["compression"]?.toBoolean() ?: false
+        }
     }
 
     override fun isSessionOpen(): Boolean = sessionOpen
@@ -961,9 +972,7 @@ class SSH : AbsTransport, ConnectionMonitor, InteractiveCallback, AuthAgentCallb
 
     override fun getPortForwards(): List<PortForward> = portForwards
 
-    override fun addPortForward(portForward: PortForward): Boolean {
-        return portForwards.add(portForward)
-    }
+    override fun addPortForward(portForward: PortForward): Boolean = portForwards.add(portForward)
 
     override fun removePortForward(portForward: PortForward): Boolean {
         // Make sure we don't have a phantom forwarder.
@@ -977,8 +986,9 @@ class SSH : AbsTransport, ConnectionMonitor, InteractiveCallback, AuthAgentCallb
             return false
         }
 
-        if (!authenticated)
+        if (!authenticated) {
             return false
+        }
 
         return when (portForward.type) {
             HostConstants.PORTFORWARD_LOCAL -> {
@@ -1044,8 +1054,9 @@ class SSH : AbsTransport, ConnectionMonitor, InteractiveCallback, AuthAgentCallb
             return false
         }
 
-        if (!authenticated)
+        if (!authenticated) {
             return false
+        }
 
         return when (portForward.type) {
             HostConstants.PORTFORWARD_LOCAL -> {
@@ -1109,12 +1120,10 @@ class SSH : AbsTransport, ConnectionMonitor, InteractiveCallback, AuthAgentCallb
 
     override fun getDefaultPort(): Int = DEFAULT_PORT
 
-    override fun getDefaultNickname(username: String?, hostname: String?, port: Int): String {
-        return if (port == DEFAULT_PORT) {
-            String.format(Locale.US, "%s@%s", username, hostname)
-        } else {
-            String.format(Locale.US, "%s@%s:%d", username, hostname, port)
-        }
+    override fun getDefaultNickname(username: String?, hostname: String?, port: Int): String = if (port == DEFAULT_PORT) {
+        String.format(Locale.US, "%s@%s", username, hostname)
+    } else {
+        String.format(Locale.US, "%s@%s:%d", username, hostname, port)
     }
 
     /**
@@ -1140,8 +1149,9 @@ class SSH : AbsTransport, ConnectionMonitor, InteractiveCallback, AuthAgentCallb
         val hostname = uri.host
         val username = uri.userInfo
         var port = uri.port
-        if (port < 0)
+        if (port < 0) {
             port = DEFAULT_PORT
+        }
         val nickname = getDefaultNickname(username, hostname, port)
 
         return Host.createSshHost(
@@ -1158,8 +1168,9 @@ class SSH : AbsTransport, ConnectionMonitor, InteractiveCallback, AuthAgentCallb
         selection[HostConstants.FIELD_HOST_HOSTNAME] = uri.host ?: ""
 
         var port = uri.port
-        if (port < 0)
+        if (port < 0) {
             port = DEFAULT_PORT
+        }
         selection[HostConstants.FIELD_HOST_PORT] = port.toString()
         selection[HostConstants.FIELD_HOST_USERNAME] = uri.userInfo ?: ""
     }
@@ -1184,14 +1195,17 @@ class SSH : AbsTransport, ConnectionMonitor, InteractiveCallback, AuthAgentCallb
                         val pubkey = pair.public as RSAPublicKey
                         pubKeys[entry.key] = RSASHA1Verify.get().encodePublicKey(pubkey)
                     }
+
                     is DSAPrivateKey -> {
                         val pubkey = pair.public as DSAPublicKey
                         pubKeys[entry.key] = DSASHA1Verify.get().encodePublicKey(pubkey)
                     }
+
                     is ECPrivateKey -> {
                         val pubkey = pair.public as ECPublicKey
                         pubKeys[entry.key] = ECDSASHA2Verify.getVerifierForKey(pubkey).encodePublicKey(pubkey)
                     }
+
                     is Ed25519PrivateKey -> {
                         val pubkey = pair.public as Ed25519PublicKey
                         pubKeys[entry.key] = Ed25519Verify.get().encodePublicKey(pubkey)
@@ -1213,8 +1227,9 @@ class SSH : AbsTransport, ConnectionMonitor, InteractiveCallback, AuthAgentCallb
         }
         if (useAuthAgent == HostConstants.AUTHAGENT_CONFIRM) {
             val holder = manager?.loadedKeypairs?.get(nickname)
-            if (holder != null && holder.pubkey?.confirmation == true && !promptForPubkeyUse(nickname))
+            if (holder != null && holder.pubkey?.confirmation == true && !promptForPubkeyUse(nickname)) {
                 return null
+            }
         }
         return manager?.getKey(nickname)
     }
@@ -1253,25 +1268,26 @@ class SSH : AbsTransport, ConnectionMonitor, InteractiveCallback, AuthAgentCallb
         return true
     }
 
-    override fun removeIdentity(publicKey: ByteArray): Boolean {
-        return manager?.removeKey(publicKey) ?: false
-    }
+    override fun removeIdentity(publicKey: ByteArray): Boolean = manager?.removeKey(publicKey) ?: false
 
     override fun isAgentLocked(): Boolean = agentLockPassphrase != null
 
     override fun requestAgentUnlock(unlockPassphrase: String): Boolean {
-        if (agentLockPassphrase == null)
+        if (agentLockPassphrase == null) {
             return false
+        }
 
-        if (agentLockPassphrase == unlockPassphrase)
+        if (agentLockPassphrase == unlockPassphrase) {
             agentLockPassphrase = null
+        }
 
         return agentLockPassphrase == null
     }
 
     override fun setAgentLock(lockPassphrase: String): Boolean {
-        if (agentLockPassphrase != null)
+        if (agentLockPassphrase != null) {
             return false
+        }
 
         agentLockPassphrase = lockPassphrase
         return true
@@ -1299,10 +1315,12 @@ class SSH : AbsTransport, ConnectionMonitor, InteractiveCallback, AuthAgentCallb
             Pattern.CASE_INSENSITIVE
         )
 
-        private const val conditions = (ChannelCondition.STDOUT_DATA
+        private const val conditions = (
+            ChannelCondition.STDOUT_DATA
                 or ChannelCondition.STDERR_DATA
                 or ChannelCondition.CLOSED
-                or ChannelCondition.EOF)
+                or ChannelCondition.EOF
+            )
 
         @JvmStatic
         fun getProtocolName(): String = PROTOCOL
@@ -1311,8 +1329,9 @@ class SSH : AbsTransport, ConnectionMonitor, InteractiveCallback, AuthAgentCallb
         fun getUri(input: String): Uri? {
             val matcher = hostmask.matcher(input)
 
-            if (!matcher.matches())
+            if (!matcher.matches()) {
                 return null
+            }
 
             val sb = StringBuilder()
 
@@ -1347,13 +1366,11 @@ class SSH : AbsTransport, ConnectionMonitor, InteractiveCallback, AuthAgentCallb
         }
 
         @JvmStatic
-        fun getFormatHint(context: Context): String {
-            return String.format(
-                "%s@%s:%s",
-                context.getString(R.string.format_username),
-                context.getString(R.string.format_hostname),
-                context.getString(R.string.format_port)
-            )
-        }
+        fun getFormatHint(context: Context): String = String.format(
+            "%s@%s:%s",
+            context.getString(R.string.format_username),
+            context.getString(R.string.format_hostname),
+            context.getString(R.string.format_port)
+        )
     }
 }

--- a/app/src/main/res/values/strings.xml
+++ b/app/src/main/res/values/strings.xml
@@ -684,6 +684,8 @@
 	<string name="terminal_auth_pubkey_specific">"Attempting 'publickey' authentication with a specific public key"</string>
 	<!-- Message shown in terminal when requesting biometric authentication to unlock a key. Parameter is the key nickname. -->
 	<string name="terminal_auth_biometric">"Requesting biometric authentication for key '%1$s'"</string>
+	<!-- Message shown when a biometric key has been invalidated due to new fingerprint enrollment. Parameter is the key nickname. -->
+	<string name="terminal_auth_biometric_invalidated">"Biometric key '%1$s' has been invalidated. This can happen when new fingerprints are enrolled on the device. Please generate a new key."</string>
 	<!-- Message shown in terminal when public key authentication fails. Parameter is the key nickname. -->
 	<string name="terminal_auth_pubkey_fail">"Authentication method 'publickey' with key '%1$s' failed"</string>
 

